src/Http/Routing/src/Matching/HeaderMatcherPolicy.cs

```diff
@@ -0,0 +1,180 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Routing.Matching;
+using Microsoft.Extensions.Options;
+
+namespace Microsoft.AspNetCore.Routing
+{
+    public sealed class HeaderMatcherPolicy : MatcherPolicy, IEndpointComparerPolicy, IEndpointSelectorPolicy
+    {
+        private readonly IOptionsMonitor<HeaderMatcherPolicyOptions> _options;
+
+        public HeaderMatcherPolicy(IOptionsMonitor<HeaderMatcherPolicyOptions> options)
+        {
+            _options = options ?? throw new ArgumentNullException(nameof(options));
+        }
+
+        /// <inheritdoc/>
+        // Run after HttpMethodMatcherPolicy (-1000) and HostMatcherPolicy (-100), but before default (0)
+        public override int Order => -50;
+
+        /// <inheritdoc/>
+        public IComparer<Endpoint> Comparer => new HeaderMetadataEndpointComparer();
+
+        /// <inheritdoc/>
+        bool IEndpointSelectorPolicy.AppliesToEndpoints(IReadOnlyList<Endpoint> endpoints)
+        {
+            if (endpoints == null)
+            {
+                throw new ArgumentNullException(nameof(endpoints));
+            }
+
+            // When the node contains dynamic endpoints we can't make any assumptions.
+            if (MatcherPolicy.ContainsDynamicEndpoints(endpoints))
+            {
+                return true;
+            }
+
+            return AppliesToEndpointsCore(endpoints);
+        }
+
+        /// <inheritdoc/>
+        public Task ApplyAsync(HttpContext httpContext, CandidateSet candidates)
+        {
+            if (httpContext == null)
+            {
+                throw new ArgumentNullException(nameof(httpContext));
+            }
+
+            if (candidates == null)
+            {
+                throw new ArgumentNullException(nameof(candidates));
+            }
+
+            int maxRequestHeadersToInspect = _options.CurrentValue.MaximumRequestHeaderValuesToInspect;
```

Comment:
Why did you choose to make `MaximumRequestHeaderValuesToInspect` a global setting and not something configured on `IHeaderMetadata`?